### PR TITLE
feat(dev-infra): create a wizard for building commit messages

### PR DESCRIPTION
**Pull request by josephperrott**

### CODE REVIEW

**Comment by sonukapoor** [comment]
I was just thinking today that such a feature would be super useful. 👍

File: dev-infra/commit-message/config.ts [hunk 1]
```diff
@@ -39,36 +39,66 @@ export enum ScopeRequirement {
...
     scope: ScopeRequirement.Required,
   },
   release: {
+    name: 'release - A release point in the repository',
+    value: 'release',
+    short: 'release',
     scope: ScopeRequirement.Forbidden,
   },
   test: {
+    name: 'test - Adds tests to better coverage missing test cases or improve testing method',
```
Comment (devversion) [inline dev-infra/commit-message/config.ts:99]: Would this read better like the following? I added `corrections` as that matches with what we have in `CONTRIBUTIONS.md`

```suggestion
    name: 'test - Improvements or corrections made to the project's test suite',
```

File: dev-infra/commit-message/wizard.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import {writeFileSync} from 'fs';
+
+import {info} from '../utils/console';
+
+import {buildCommitMessage} from './builder';
+
+const defaultCommitMessage = `<type>(<scope>): <summary>
+
+# <Describe the motivation behind this change - explain WHY you are making this change. Wrap all
+#  lines at 100 characters.>\n\n`;
```
Comment (devversion) [inline dev-infra/commit-message/wizard.ts:17]: Neither this comment is wrapped at 100 characters! 😄 Can we wrap it as a good example and for increased readability?

File: dev-infra/utils/inquirer-autocomplete-typings.d.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+declare module "inquirer-autocomplete-prompt" {
```
Comment (devversion) [inline dev-infra/utils/inquirer-autocomplete-typings.d.ts:9]: Add a comment saying that `inquirer-autocomplete-prompt` does not have any types? Potentially also create an issue upstream asking for types being shipped (even if it is just that)? https://github.com/mokkabonna/inquirer-autocomplete-prompt/issues/61 has been closed unfortunately.

File: dev-infra/utils/console.ts [hunk 4]
```diff
@@ -108,3 +152,18 @@ function getLogLevel() {
...
+
+
+/** Force StdIn to be available, exiting as an error if it is unable to be made available. */
+export function forceStdInToBeAvailable() {
+  if (process.stdin.isTTY) {
+    debug('STDIN already available in tty, skipping force');
+    return;
+  }
+  try {
+    forceStdinTty();
```
Comment (devversion) [inline dev-infra/utils/console.ts:164]: Unfortunately this does not seem to work on Windows as `/dev/tty` is not a thing there.

* https://github.com/dosentmatter/force-stdin-tty/blob/master/src/index.ts
* https://github.com/dosentmatter/force-stdin-tty/issues/2#issuecomment-629391725

File: dev-infra/utils/console.ts [hunk 5]
```diff
@@ -26,6 +31,45 @@ export async function promptConfirm(message: string, defaultValue = false): Prom
       .result;
 }
 
+/** Prompts the user to select an option from a filterable autocomple list. */
+export async function promptAutocomplete(
```
Comment (devversion) [inline dev-infra/utils/console.ts:35]: Nice, love the use of an overload. Can you please add a separate comment for the overload below (capturing the specific difference too?)

File: dev-infra/utils/console.ts [hunk 6]
```diff
@@ -7,7 +7,12 @@
  */
 
 import chalk from 'chalk';
-import {prompt} from 'inquirer';
+import {forceStdinTty} from 'force-stdin-tty';
+import {ListChoiceOptions, prompt, registerPrompt} from 'inquirer';
+import * as inquirerAutocomplete from 'inquirer-autocomplete-prompt';
+
+// Registers the autocomplete prompt type for inquirer.
+registerPrompt('autocomplete', inquirerAutocomplete);
```
Comment (devversion) [inline dev-infra/utils/console.ts:15]: Might be worth considering registering this lazily (if it works in inqurier).

File: dev-infra/utils/console.ts [hunk 7]
```diff
@@ -26,6 +31,45 @@ export async function promptConfirm(message: string, defaultValue = false): Prom
...
+                   message,
+                   source: (_: any, input: string) => {
+                     if (!input) {
+                       return Promise.resolve(choices);
+                     }
+                     return Promise.resolve(choices.filter(scope => {
+                       if (typeof scope === 'string') {
+                         return scope.includes(input);
+                       }
+                       return scope.name!.includes(input);
```
Comment (devversion) [inline dev-infra/utils/console.ts:58]: Is this leaking logic specific to the commit-message wizard? e.g. `scope` and `scope.name`. I think if this function exists in `utils` it should be more generic.

File: dev-infra/utils/console.ts [hunk 8]
```diff
@@ -26,6 +31,45 @@ export async function promptConfirm(message: string, defaultValue = false): Prom
...
+export async function promptAutocomplete(
+    message: string, choices: (string|ListChoiceOptions)[],
+    noChoiceText?: string): Promise<string|false>;
+export async function promptAutocomplete(
+    message: string, choices: (string|ListChoiceOptions)[],
+    noChoiceText?: string): Promise<string|false> {
+  if (noChoiceText) {
+    choices = [noChoiceText, ...choices];
+  }
+  const result = (await (prompt as any)({
```
Comment (devversion) [inline dev-infra/utils/console.ts:46]: Can we have a comment on why we cast to `any`. I'm sure it's because of missing typings, but maybe link here to an issue upstream too and explain?

File: dev-infra/commit-message/wizard.ts [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  // Set the default commit message to be updated if the user cancels out of the wizard in progress
+  writeFileSync(args.filePath, defaultCommitMessage);
+
+  /** The generated commit message. */
+  const commitMessage = await buildCommitMessage();
+
+  writeFileSync(args.filePath, commitMessage);
+  process.exitCode = 0;
```
Comment (devversion) [inline dev-infra/commit-message/wizard.ts:36]: Wouldn't the script exit automatically with exit code zero if there are no errors? Why is this needed?

File: dev-infra/commit-message/wizard.ts (hunk 2 above)
Comment (devversion) [reply dev-infra/commit-message/wizard.ts:17]: Also, this is a great comment. Should it be in the built commit message too? I might be wrong; but right now just the header is set but it's not made clear that a description of changes needs to be added & where?

I assume we could append it to the built commit message (only the comments)

**Review by devversion — COMMENTED** [review]
🎉

File: dev-infra/commit-message/builder.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+/** Prompts in the terminal for the commit message's type. */
+async function promptForCommitMessageType(): Promise<CommitType> {
+  info('The type of change in the commit, Allows a reader to know the effect of the change,');
+  info('whether it brings about something new, adds additional testing, documents the `project,');
+  info('etc.');
+
+  /** The key of a commit message type, selected by the user via prompt. */
+  const typeKey =
```
[the remark this answers is not in the document]
Comment (josephperrott) [reply dev-infra/commit-message/builder.ts:37]: Yeah, we can have it be `typeName`.  Though in this case it is both the name and the key.

File: dev-infra/commit-message/wizard.ts (hunk 2 above)
Comment (devversion) [reply dev-infra/commit-message/wizard.ts:21]: Hopla!! 😅 I didn't realize this is a template string and joined the line below.

File: dev-infra/utils/console.ts [hunk 11]
```diff
@@ -108,3 +157,21 @@ function getLogLevel() {
...
+
+
+/** Force StdIn to be available, exiting as an error if it is unable to be made available. */
+export function forceStdInToBeAvailable() {
+  if (process.stdin.isTTY) {
+    debug('STDIN already available in tty, skipping force');
+    return;
+  }
+  if (platform() === 'win32') {
+    debug(`STDIN can't be forced as STDIN doesn't exist on windows OS.`);
```
Comment (devversion) [inline dev-infra/utils/console.ts:169]: What are the consequences for Windows users? Is the stdin not available by default in husky git hooks?

File: dev-infra/commit-message/config.ts [hunk 12]
```diff
@@ -39,36 +39,66 @@ export enum ScopeRequirement {
 
 /** A commit type */
 export interface CommitType {
+  value: string;
+  name: string;
+  short: string;
```
Comment (devversion) [inline dev-infra/commit-message/config.ts:44]: Just noticed: Why is the `short` and `name` property needed here? I remember seeing a discussion with @gkalpak that the name/value could be inferred from the key in the `COMMIT_TYPES` record?

File: dev-infra/commit-message/wizard.ts [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {writeFileSync} from 'fs';
+
+import {info} from '../utils/console';
+
+import {buildCommitMessage} from './builder';
+
+/** The source triggering the git commit message creation. */
```
Comment (devversion) [inline dev-infra/commit-message/wizard.ts:14]: Optional Nit when you come across changing other things: Also add the link `https://git-scm.com/docs/githooks#_prepare_commit_msg` to the JSDoc. That would help with readability as this type is most local to where the `source` is dealt with.

File: dev-infra/commit-message/builder.ts [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    info(`Skipping scope selection as the '${type.value}' type does not allow scopes`);
+    return false;
+  }
+
+  /** Commit message configuration */
+  const config = getCommitMessageConfig();
+
+  /* The commit message scope*/
+  info('The area of the repository the changes in this commit most affects.');
+  /** The scope of the commit, selected by the user via prompt. */
```
Comment (devversion) [inline dev-infra/commit-message/builder.ts:56]: The return value of these functions should be already explained in the function description, right?

File: dev-infra/commit-message/builder.ts [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+  // If the commit type's scope requirement is forbidden, return early.
+  if (type.scope === ScopeRequirement.Forbidden) {
+    info(`Skipping scope selection as the '${type.value}' type does not allow scopes`);
+    return false;
+  }
+
+  /** Commit message configuration */
+  const config = getCommitMessageConfig();
+
+  /* The commit message scope*/
```
Comment (devversion) [inline dev-infra/commit-message/builder.ts:54]: I assume this comment should be removed? Also it's not describing any declaration, so a multi-line comment is not needed. Same for the `return` values below. I'm thinking a multi-line comment (that is captured in IDEs and JSDoc) should only be used for declarations where it has an effect. Like you did above with the `if` check.

**Review by devversion — APPROVED** [review]
LGTM a few minor cleanup comments

File: dev-infra/utils/console.ts (hunk 11 above)
Comment (josephperrott) [reply dev-infra/utils/console.ts:169]: Could you check for me to confirm?  But I believe that STDIN is available on windows on husky runs.

File: dev-infra/commit-message/config.ts (hunk 12 above)
Comment (devversion) [reply dev-infra/commit-message/config.ts]: Wouldn't it be better to generate these options on the fly? I think the constants should be only describing the type of commit. There could still be a property like `description` but the rest could be generated lazily.

File: dev-infra/utils/console.ts (hunk 11 above)
Comment (devversion) [reply dev-infra/utils/console.ts:169]: I'm getting `Unable to redirect terminal input back into node process`. Should we first check before running the force stdin?

File: dev-infra/commit-message/config.ts (hunk 12 above)
Comment (josephperrott) [reply dev-infra/commit-message/config.ts]: Done

Comment (devversion) [reply dev-infra/commit-message/config.ts]: thanks! looks so great 🎉
